![numpy](https://raw.githubusercontent.com/donnemartin/data-science-ipython-notebooks/master/images/numpy.png)

[NumPy](https://www.numpy.org/) is the fundamental package for scientific computing with Python. Many other data science packages, especially those that work with matrices, rely on it for its speed and utility.

<img src="images/numpy.png">

<img src="images/numpyuse1.png">

<img src="images/numpyuse2.png">

How most people work with Numpy
<ol><li>Not at ALL</li>
    <li>Much work done using Numpy as back-bone</li>
    <li>Work with tools that need Numpy input or output Numpy arrays</li>
</ol>

# Objectives

- Use NumPy to create arrays and perform efficient operations with them
- Use NumPy's other mathematical tools relevant to data analysis

In [2]:
import numpy as np

For numpy, the standard alias is `np`.

In [ ]:
x = np.array([1, 2, 3])

# NumPy Arrays

Python lists and NumPy arrays can both hold numbers. 
- However, Python lists have limited functionality for mathematical operations. 
- NumPy arrays make it easy and fast to do math with a collection of numbers.

In [3]:
y = [4,5,6]  # Python List
x = np.array([1, 2, 3])  # Numpy Array
print(y)
print(x)
print(type(y))
print(type(x))

[4, 5, 6]
[1 2 3]
<class 'list'>
<class 'numpy.ndarray'>


Note that there is an [`array` class in base Python](https://docs.python.org/3/library/array.html), but we will not be using it. It is essentially a list constrained to one type (e.g. `int` ("i" below)).

In [4]:
import array
x = array.array('i', [1, 2, 3])
print(x)
print(type(x))

array('i', [1, 2, 3])
<class 'array.array'>


## Character Arrays

NumPy arrays can either hold strings or numbers, not both. There is a special `chararray` object you can use with strings.

In [5]:
names_list = ['Bob', 'John', 'Sally']

# Use numpy.array for numbers and numpy.char.array for strings.

names_array = np.char.array(['Bob', 'John', 'Sally'])

print(names_list)
print(names_array)
type(names_array)

['Bob', 'John', 'Sally']
['Bob' 'John' 'Sally']


numpy.chararray

Note what happens when we try to put both data types into a single array:

In [6]:
np.array([4, 'bob'])

array(['4', 'bob'], dtype='<U21')

In [20]:
# The character array has string-functionality that numeric
# arrays don't have.
print(names_array)
names_array.endswith('b')

['Bob' 'John' 'Sally']


array([ True, False, False])

Numpy String Operations 
https://numpy.org/doc/stable/reference/routines.char.html

## Numeric Arrays

Let's make a list and an array of three numbers, and see how they function differently.

In [7]:
numbers_list = [0, 5, 7]
numbers_array = np.array([0, 5, 7])

### Type

Numeric arrays are have the `ndarray` type, which is short for N-Dimensional Array.

In [8]:
print(type(numbers_list))
print(type(numbers_array))

<class 'list'>
<class 'numpy.ndarray'>


## Arithmetic Operations

Arithmetic operators (e.g. +, -, \*, and /) work according to mathematical principles for arrays, unlike with lists. These operations are done "element-wise".

In [10]:
# multiply the array by 3
print(numbers_array)
numbers_array * 3

[0 5 7]


array([ 0, 15, 21])

In [11]:
# multiply the list by 3

numbers_list * 3

[0, 5, 7, 0, 5, 7, 0, 5, 7]

In [11]:
numbers_array + 20

array([20, 25, 27])

In [12]:
numbers_list + 20

TypeError: can only concatenate list (not "int") to list

### Speed

Below, you will find a piece of code we will use to compare the speed of operations on lists vs arrays.

In [13]:
size_of_vec = 1000
# Python Lists
XL = list(range(size_of_vec))
YL = list(range(size_of_vec))

# Numpy Arrays
XA = np.array(range(size_of_vec))
YA = np.array(range(size_of_vec))

In [14]:
%timeit [XL[i] + YL[i] for i in range(size_of_vec)]

81.2 µs ± 1.43 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [15]:
%timeit XA + YA

646 ns ± 1.23 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


## Array Attributes and Methods

Type `numbers_list.` and then hit `TAB`. What options do you have?

In [ ]:
numbers_list.

The names of standard Python list methods appear:

- `append(x)` (add x to the end of the list)
- `clear()` (delete all elements of the list)
- `copy()` (make a copy of the list)
- `count(x)` (return the number of instances of x in the list)
- `extend([x, y])` (add x and y to the end of the list)
- `index(x)` (return the position in the list of x)
- `insert(x, y)` (insert y into position x in the list)
- `pop(i=-1)` (remove and return the element at position i in the list)
- `remove(x)` (remove x from the list)
- `reverse()` (reverse the order of the elements of the list)
- `sort()` (sort the elements of the list)

Now type `numbers_array.` and then hit `TAB`. What options do you have?

In [ ]:
numbers_array.

Turns out, there are a _bunch_ of new tools!

### Numeric Methods

- `max()` (return the greatest value in the array)

In [27]:
print(numbers_array)
numbers_array.max()

[0 5 7]


7

- `mean()` (return the arithmetic mean of the array)

In [16]:
numbers_array.mean()

4.0

- `min()` (return the smallest value in the array)

In [17]:
numbers_array.min()

0

- `round()` (round each entry in the array to a specified number of decimal places)

In [18]:
np.array([9.5, 1.2, 6.3]).round()

array([10.,  1.,  6.])

- `std()` (return the standard deviation of the array)

In [19]:
numbers_array.std()

2.943920288775949

- `sum()` (return the sum of the array's elements)

In [20]:
numbers_array.sum()

12

### Boolean Methods

- `all()` (returns True iff bool(element) == True for all elements in the array)

In [21]:
print(numbers_array)
numbers_array.all()

[0 5 7]


False

- `any()` (returns True if bool(element) == True for some element in the array)

In [22]:
numbers_array.any()

True

## Multi-Dimensional Indexing

Arrays are especially powerful when dealing with numbers in multiple dimensions - for example, a datset with rows and columns. We will primarily work with such 2-dimensional arrays.

<img src="images/md.png" style="width:600px;height:500px">

In [23]:
nums = np.array([[1, 2, 3], [4, 5, 6]])
print(nums)

[[1 2 3]
 [4 5 6]]


In [24]:
print(nums.shape)

(2, 3)


In [6]:
nums[0, 2]

3

In [25]:
%timeit nums[0, 2]

93.3 ns ± 0.785 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [26]:
%timeit nums[0][2]

183 ns ± 1.45 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


### Slicing

Use a trailing comma or colon to [slice](https://numpy.org/doc/stable/reference/arrays.indexing.html) sections of an array

<img src="images/numpar4.png" style="width:500px;height:500px">


In [27]:
x = np.array([[1,2,3,4], [5,6,7,8], [9,10,11,12],[13,14,15,16]])
print(x[1,2:4])

[7 8]


<img src="images/numpar3.jpg">

In [20]:
print(x)
x[1]

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]


array([5, 6, 7, 8])

In [28]:
x[1,]

array([5, 6, 7, 8])

In [29]:
x[1, :]

array([5, 6, 7, 8])

In [39]:
x[:, 1]
y = 2
z=5
if (y == 2 and z == 5):
    print(y)

a = (y == 2 and z == 5)
print(a)
 
b = (y == 2 & z == 5)
print(b)

2
True
False



### Reshaping

In [30]:
new = np.array([[1, 2, 3], [4, 5, 6]])
new

array([[1, 2, 3],
       [4, 5, 6]])

- `shape` (stores the dimension of the array)

In [31]:
#  .shape  returns  (rows,columns)
new.shape

(2, 3)

- `ravel()` (reduce array to one dimension)

In [32]:
new.ravel()

array([1, 2, 3, 4, 5, 6])

- `reshape()` (return an array with the specified dimensions)

In [35]:
print(new)
new.reshape(3, 2)

[[1 2 3]
 [4 5 6]]


array([[1, 2],
       [3, 4],
       [5, 6]])

In [46]:
x.reshape(-1,1)

array([[ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12],
       [13],
       [14],
       [15],
       [16]])

# NumPy Functions

NumPy has a bunch of functions, besides array methods, that are helpful for working with arrays.

## Array Constructors

In [47]:
print(np.zeros(10))
print(np.ones(10))
print(np.arange(10, dtype=float))
print(np.linspace(0.1, 1, 10))

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
[0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]


### `np.concatenate()`

In [48]:
np.concatenate([[1, 2], [3, 4]])

array([1, 2, 3, 4])

### `np.hstack()`

In [37]:
arr1 = np.array([[1, 3], [5, 7], [9, 11]])
arr2 = np.array([[2, 4], [6, 8], [10, 12]])

np.hstack((arr1, arr2))

array([[ 1,  3,  2,  4],
       [ 5,  7,  6,  8],
       [ 9, 11, 10, 12]])

### `np.vstack()`

In [39]:
print(arr1)
print(arr2)
np.vstack((arr1, arr2))

[[ 1  3]
 [ 5  7]
 [ 9 11]]
[[ 2  4]
 [ 6  8]
 [10 12]]


array([[ 1,  3],
       [ 5,  7],
       [ 9, 11],
       [ 2,  4],
       [ 6,  8],
       [10, 12]])

## Filtering

In [40]:
data = np.array([10, 3, 4, 7, 6])

In [41]:
data < 5

array([False,  True,  True, False, False])

In [42]:
data[data < 5]

array([3, 4])

### `np.where()`

In [44]:
print(data)
np.where(data < 5, "Low", "High")

[10  3  4  7  6]


array(['High', 'Low', 'Low', 'High', 'High'], dtype='<U4')

### `np.select()`

In [45]:
conditions = [data < 5, data > 9]

choices = ['small', 'big!']

In [46]:
np.select(conditions, choices, default='other')

array(['big!', 'small', 'small', 'other', 'other'], dtype='<U5')

# Broadcasting

Two arrays can be combined with mathematical operations via [broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html).  

From the [docs](https://numpy.org/doc/stable/user/basics.broadcasting.html):

    When operating on two arrays, NumPy compares their shapes element-wise. It starts with the 
    trailing dimensions and works its way forward. Two dimensions are compatible when
    - they are equal, or
    - one of them is 1
    
Let's try to figure out what will happen in the operations below

In [47]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([10, 100, 1000])
print(arr1)
print(arr2)

[[1 2 3]
 [4 5 6]]
[  10  100 1000]


In [52]:
print(arr1.shape)
print(arr2.shape)

(2, 3)
(3,)


In [48]:
print(arr1)
print(arr2)

arr1 * arr2

[[1 2 3]
 [4 5 6]]
[  10  100 1000]


array([[  10,  200, 3000],
       [  40,  500, 6000]])

In [49]:
arr3 = np.array([10, 100])
arr3.shape

(2,)

In [50]:
arr1 * arr3

ValueError: operands could not be broadcast together with shapes (2,3) (2,) 

In [53]:
arr5 = np.array([[10],[100]])
arr5.shape

(2, 1)

In [54]:
arr1 * arr5

array([[ 10,  20,  30],
       [400, 500, 600]])

# Matrix Multiplication

In [55]:
np.linalg

<module 'numpy.linalg' from '/Users/joecomeausx/opt/anaconda3/lib/python3.9/site-packages/numpy/linalg/__init__.py'>

The linear algebra module contains many good tools for working with matrices.

Let's use the random module to make two random arrays:

In [56]:
np.random.seed(42)
X = np.random.rand(10, 5)
Y = np.random.rand(5, 5)
X, Y

(array([[0.37454012, 0.95071431, 0.73199394, 0.59865848, 0.15601864],
        [0.15599452, 0.05808361, 0.86617615, 0.60111501, 0.70807258],
        [0.02058449, 0.96990985, 0.83244264, 0.21233911, 0.18182497],
        [0.18340451, 0.30424224, 0.52475643, 0.43194502, 0.29122914],
        [0.61185289, 0.13949386, 0.29214465, 0.36636184, 0.45606998],
        [0.78517596, 0.19967378, 0.51423444, 0.59241457, 0.04645041],
        [0.60754485, 0.17052412, 0.06505159, 0.94888554, 0.96563203],
        [0.80839735, 0.30461377, 0.09767211, 0.68423303, 0.44015249],
        [0.12203823, 0.49517691, 0.03438852, 0.9093204 , 0.25877998],
        [0.66252228, 0.31171108, 0.52006802, 0.54671028, 0.18485446]]),
 array([[0.96958463, 0.77513282, 0.93949894, 0.89482735, 0.59789998],
        [0.92187424, 0.0884925 , 0.19598286, 0.04522729, 0.32533033],
        [0.38867729, 0.27134903, 0.82873751, 0.35675333, 0.28093451],
        [0.54269608, 0.14092422, 0.80219698, 0.07455064, 0.98688694],
        [0.7722447

We can multiply these matrices together, according to the rule of dot-multiplication:

$\begin{equation}
\begin{bmatrix}
\vec{a_1} \\
... \\
\vec{a_r}
\end{bmatrix}
\times
\begin{bmatrix}
\vec{b_1} & ... & \vec{b_c}
\end{bmatrix}
=
\begin{bmatrix}
\vec{a_1}\cdot\vec{b_1} & ... & \vec{a_1}\cdot\vec{b_c} \\
... & ... & ... \\
\vec{a_r}\cdot\vec{b_1} & ... & \vec{a_r}\cdot\vec{b_c}
\end{bmatrix}
\end{equation}$


In [62]:
X.dot(Y)

array([[1.96947098, 0.68844411, 1.62593817, 0.81114581, 1.43996725],
       [1.41448678, 0.58650929, 1.36189545, 1.07344574, 1.44924311],
       [1.49329385, 0.39372317, 1.07064311, 0.52336391, 0.89978943],
       [1.12157524, 0.43022148, 1.01493314, 0.63477194, 0.98819733],
       [1.38640974, 0.70814211, 1.14069762, 1.0572552 , 1.17721607],
       [1.50261206, 0.8585383 , 1.67845967, 0.97712588, 1.29636284],
       [2.03221277, 0.82927694, 1.42464369, 1.43274317, 2.0560098 ],
       [1.8138242 , 0.8639648 , 1.45145232, 1.18193516, 1.59626598],
       [1.28150938, 0.32731549, 0.97108313, 0.42268239, 1.3240406 ],
       [1.57131887, 0.79602502, 1.55411901, 0.98397619, 1.31384314]])

# Level Up: Other NumPy Tools

NumPy comes with an assortment of mathematical tools that can come in handy, separate from arrays.

## Trigonometry 

- `np.pi` for $\pi$

In [ ]:
np.pi

- `np.sin()` for the sine function

In [ ]:
np.sin(np.pi / 6)

## Sequences


- `np.cumsum()` to calculate, recursively, the sum of sequence terms

In [ ]:
np.cumsum([1, 4, 9, 16])

- `np.diff()` to calculate, recursively, the differences between sequence terms

In [ ]:
np.diff([1, 4, 9, 16])

## Logarithms
- `np.exp()` for Euler's number with exponent

In [ ]:
np.exp(2)

- `np.log()` for logarithms

In [ ]:
np.log(10)

## np.nan

NaN stands for "not a number" - NumPy's `nan` class is a handy way of representing these.  These are very useful for representing missing data.

Since `np.nan` is is a float, we don't get errors when doing operations on arrays that have missing data. 

In [ ]:
type(np.nan)

In [ ]:
arr5 = np.array([1, 10, np.nan])

In [ ]:
arr5.mean()

Even though the array has a NaN, we don't get an error in calculating its mean. Moreover, we can do this:

In [ ]:
np.nansum(arr5) / len(arr5)

Is the right measure of the mean? Well, maybe. But if not, we also have this:

In [ ]:
np.nanmean(arr5)

## np.inf

Sometimes you will end up with values of infinity represented by `np.inf`, such as if you divide by zero. `np.inf` is a float, like `np.nan` is.

In [ ]:
np.array([2])/np.array([0])

In [ ]:
type(np.inf)

In [ ]:
np.isfinite(np.inf)

This can also be useful when handling edge cases in custom functions

In [ ]:
def inv(x):
    return x**(-2)

In [ ]:
inv(0)

In [ ]:
def inverse(x):
    if x == 0:
        val = np.inf
    else:
        val = x**(-2)
    return val

In [ ]:
inverse(0)

# Level Up: Matrix Decompositions

We can do eigendecompositions and singular value decompositions:

## Eigendecomposition

In [ ]:
diagonalize = np.linalg.eig(Y)
eigvals, eigvecs = diagonalize
eigvals, eigvecs

In [ ]:
eigvecs.dot(np.diag(eigvals)).dot(np.linalg.inv(eigvecs))

In [ ]:
Y

### `.norm()`

In [ ]:
np.linalg.norm(eigvecs[:, 0])

## Singular Value Decomposition

In [ ]:
svd = np.linalg.svd(X)
leftvecs, singvals, rightvecs = svd
leftvecs, singvals, rightvecs

In [ ]:
diag = np.diag(singvals)
pdiag = np.vstack((diag, np.zeros((5, 5))))
leftvecs.dot(pdiag).dot(rightvecs)

In [ ]:
X

# Level Up: Linear Algebra

## Inverses

We can also take inverses and pseudo-inverses:

In [ ]:
np.linalg.inv(Y)

In [ ]:
np.linalg.pinv(X)

## Systems of Equations

In [ ]:
np.random.seed(42)
c = np.random.rand(5)
c

We can solve matrix equations, such as:

$Y\vec{x} = c$.

Let's plug in what we have for $Y$ and $c$:

$
\begin{bmatrix}
0.969585 & 0.775133 & 0.939499 & 0.894827 & 0.5979 \\
0.921874 & 0.0884925 & 0.195983 & 0.0452273 & 0.32533 \\
0.388677 & 0.271349 & 0.828738 & 0.356753 & 0.280935 \\
0.542696 & 0.140924 & 0.802197 & 0.0745506 & 0.986887 \\
0.772245 & 0.198716 & 0.00552212 & 0.815461 & 0.706857
\end{bmatrix}
\begin{bmatrix}
x_1 \\
x_2 \\
x_3 \\
x_4 \\
x_5
\end{bmatrix}
=
\begin{bmatrix}
0.37454 \\
0.950714 \\
0.731994 \\
0.598658 \\
0.156019
\end{bmatrix}
$

In [ ]:
x = np.linalg.solve(Y, c)
x

In [ ]:
Y.dot(x)

# Exercises

1. Write a function that will return the sum of the first $n$ terms of a 1-d array, where the user inputs both $n$ and the array.

<details>
    <summary>Answer</summary>
    <code>def first_n_sum(n, arr):
    return np.cumsum(arr)[n-1]</code> <br>
    OR <br>
    <code>def first_n_sum(n, arr):
    return np.sum(arr[:n])</code>
    </details>

2. Write a function that will return the natural logarithms of the standard deviations of the columns of an input array.

<details>
    <summary>Answer</summary>
    <code>def log_std(arr):
    return np.log(np.std(arr, axis=0))</code>
</details>    

3. How could I quickly build a 10x10 matrix of the numbers from 1 to 100?

<details>
    <summary>Answer</summary>
    <code>np.array(np.arange(1, 101)).reshape(10, 10)</code>
    </details>

4. Write a function that will return, for any input array, an array of the elements less than 0.5.

<details>
    <summary>Answer</summary>
    <code>def less_than_half(arr):
    return arr[arr < 0.5]</code>
        </details>

5. Write a function that will return, for any input number n, a 100x100 array of only n's.

<details>
    <summary>Answer</summary>
    <code>def square(n):
    return n*np.ones((100, 100))</code>
    </details>

6. Write a function that will return, for any input number n, an nxn array of only n's.

<details>
    <summary>Answer</summary>
    <code>def n_square(n):
    return n*np.ones((n, n))</code>
    </details>

7. Write a function that will take in an array, X, build a second array whose elements are doubles of X, 2X, and then return an array composed of X horizontally concatenated with 2X (X on the left and 2X on the right).

$f(X) = [X | 2X]$

<details>
    <summary>Answer</summary>
    <code>def concat_double(arr):
    return np.hstack((arr, 2*arr))</code>
</details>    

8. Write a function that will take in an array, X and two numbers m and n, and return X with an additional column that is the (pairwise) product of the mth and nth columns of X.

$f(X, m, n) = [X|X_m\times X_n]$

<details>
    <summary>Answer</summary>
    <code>def col_mult(X, m, n):
    try:
        prod = (X[:, m]*X[:, n]).reshape(-1, 1)
        return np.hstack((X, prod))
    except:
        print('m or n is too large!')</code>
    </details>

9. Write a function that will take in an array, X, and return the row of X with the smallest standard deviation.

<details>
    <summary>Answer</summary>
<code>def col_std(X):
    stds = [np.std(row) for row in X]
    smallest = stds.index(min(stds))
    return X[smallest, :]</code>
    </details>

10. Write a function that will take in an array, X, and return the row or column (input to function) of X with the largest mean.

<details>
    <summary>Answer</summary>
    <code>def largest_mean(X, row=True):
    if row:
        means = [np.mean(row) for row in X]
        largest = means.index(max(means))
        return X[largest, :]
    else:
        means = [np.mean(col) for col in X.T]
        largest = means.index(max(means))
        return X[:, largest]</code>
    </details>